Expand on rolling update recommendations, best practices (#27382)
akka/akka: akka-docs/src/main/paradox/additional/rolling-updates.md

    - If a `java.io.NotSerializableException` is thrown in `fromBinary` this is treated as a transient problem, the issue logged and the message is dropped
    - If other exceptions are thrown it can be an indication of corrupt bytes from the underlying transport, and the connection is broken
* For more zero-impact rolling upgrades, it is important to consider a strategy for serialization that can be evolved. You can find advice in
@ref:[Persistence - Schema Evolution](../persistence-schema-evolution.md), which also applies to
remote messages when deploying with rolling updates.

One approach to retiring a serializer without downtime is carried out in @ref:[two rolling upgrade steps to switch to the new serializer](../serialization.md#rolling-upgrades). 

## Cluster Sharding

During a rolling upgrade, sharded entities receiving traffic may be moved during @ref:[shard rebalancing](../cluster-sharding.md#shard-rebalancing), 
to an old or new node in the cluster, based on the pluggable allocation strategy and settings.
When an old node is stopped the shards that were running on it are moved to one of the

> other old nodes remaining in the cluster. See @ref[ClusterSingleton](#cluster-singleton) for a useful `ShardCoordinator` optimization.

Reviewer: ```suggestion
other nodes remaining in the cluster. See @ref[ClusterSingleton](#cluster-singleton) for a useful `ShardCoordinator` optimization.
```

overhead several times.

## Cluster Shutdown
 
@ref:[Coordinated Shutdown](../actors.md#coordinated-shutdown) will automatically run on SIGTERM when the cluster node sees itself as Exiting.
Thus running shutdown tasks in a JVM shutdown hook is not recommended over Coordinated Shutdown.
@ref:[Graceful shutdown](../cluster-sharding.md#graceful-shutdown) of Cluster Singletons and Cluster Sharding similarly happen automatically.
 
In case of network failures it may still be necessary to set the node's status to Down in order to complete the removal. 
@ref:[Cluster Downing](../cluster-usage.md#downing) details downing nodes and downing providers, and how using
[Split Brain Resolver](https://doc.akka.io/docs/akka-enhancements/current/split-brain-resolver.html) can help resolve
split brain during network partitions.

Additionally, [Cluster Bootstrap](https://doc.akka.io/docs/akka-management/current/bootstrap/index.html#rolling-updates)
can be leveraged during rolling updates for joining and downing nodes in the cluster.
 
## Cluster Configuration Compatibility Checks

During rolling updates the configuration from existing nodes should pass the Cluster configuration compatibility checks.

> For example, when migrating from Classic to Typed Clusters, a two step approach is possible:

Reviewer 2: Aha, accidentally found the answer while just closing another tab: it's about migrating sharded entities from untyped to typed. That should probably be clarified (see https://github.com/akka/akka/issues/26718#issuecomment-501845738).

> For example, when migrating from Classic to Typed Clusters, a two step approach is possible:

Author: That is where it is from, yes.

> For example, when migrating from Classic to Typed Clusters, a two step approach is possible:

Author: Added.

# Rolling Updates

A rolling update is the process of replacing one version of the system with another without downtime.
The changes can be new code, changed dependencies such as new Akka version, or modified configuration.

In Akka, rolling updates are typically used for a stateful Akka Cluster where you can't run two separate clusters in
parallel during the update, for example in blue green deployments.

> For rolling updates related to Akka dependency version upgrades and the migration guides, please see

Reviewer 2: Might be worth an extra section below with some basic information similar to the other ones.

Author: I would but Patrik wanted to be sure we do not duplicate content here.

@ref:[Rolling Updates and Akka versions](../project/rolling-update.md)

> #### This document covers: 

Reviewer 2: Do we need an extra TOC here?

Author: No but my thinking was if it were me, I'd find it useful to see if the doc had what I was looking for vs having to scan it, for anyone looking for their specific issues.

* [Serialization Compatibility](#serialization-compatibility)
* [Cluster Sharding](#cluster-sharding)
* [Cluster Singleton](#cluster-singleton)
* [Migrating Untyped to Typed](#migrating-untyped-to-typed)
* [Cluster Shutdown](#cluster-shutdown)
* [Cluster Configuration Compatibility Check](#cluster-configuration-compatibility-check)
 
## Serialization Compatibility

There are two parts of Akka that need careful consideration when performing an rolling update.

1. Compatibility of remote message protocols. Old nodes may send messages to new nodes and vice versa.
1. Serialization format of persisted events and snapshots. New nodes must be able to read old data, and
   during the update old nodes must be able to read data stored by new nodes.

There are many more application specific aspects for serialization changes during rolling upgrades to consider. 
For example, whether to allow dropped messages or tear down the TCP connection when the manifest is unknown.

> * When some message loss during a rolling upgrade is acceptable versus a full shutdown and restart, assuming the application recovers afterwards 

Author: Yes. That is pre-existing content.

> * When some message loss during a rolling upgrade is acceptable versus a full shutdown and restart, assuming the application recovers afterwards 

Author: https://github.com/akka/akka/pull/25595/files#diff-ab8b0bfebd460d8f00c33650c4c761daR16

> * When some message loss during a rolling upgrade is acceptable versus a full shutdown and restart, assuming the application recovers afterwards 

Reviewer 2: The bullets just feel a bit thrown together without any obvious connection to previous paragraph. Is it an explanation or a recommendation or a choice for the user? I can somehow puzzle together what is meant but if you read it for the first time, you'll probably be confused.

> * When some message loss during a rolling upgrade is acceptable versus a full shutdown and restart, assuming the application recovers afterwards 

Author: Agreed, this is now clarified and pushed.

    - If a `java.io.NotSerializableException` is thrown in `fromBinary` this is treated as a transient problem, the issue logged and the message is dropped
    - If other exceptions are thrown it can be an indication of corrupt bytes from the underlying transport, and the connection is broken
* For more zero-impact rolling upgrades, it is important to consider a strategy for serialization format that can be evolved. You can find advice in
@ref:[Persistence - Schema Evolution](../persistence-schema-evolution.md), which also applies to
remote messages when deploying with rolling updates.

> One approach to retiring a serializer without downtime is carried out in @ref:[two rolling upgrade steps to switch to the new serializer](../serialization.md#rolling-upgrades). 

Reviewer: :+1:

## Cluster Sharding

During a rolling upgrade, sharded entities receiving traffic may be moved during @ref:[shard rebalancing](../cluster-sharding.md#shard-rebalancing), 
to an old or new node in the cluster, based on the pluggable allocation strategy and settings.
When an old node is stopped the shards that were running on it may be allocated to one of the

> other old nodes remaining in the cluster. See @ref[ClusterSingleton](#cluster-singleton) for a useful `ShardCoordinator` optimization.

Author: Refers to the sentence under the singleton for `optional optimization` so that nothing is duplicated, WRT `leave the oldest node running` until last if desired.

> other old nodes remaining in the cluster. See @ref[ClusterSingleton](#cluster-singleton) for a useful `ShardCoordinator` optimization.

Author: 👍 updated thanks

There are some cases when @ref:[a full cluster restart is needed](../cluster-sharding.md#rolling-upgrades).

## Cluster Singleton

> It's more efficient to avoid moving a `ClusterSingleton` more than necessary because they typically have to recover their state

Reviewer 2: Rewording suggestion for the whole section:

> Cluster singletons are always running on the oldest node. To avoid moving cluster singletons more than necessary during a rolling upgrade, it is recommended to upgrade the oldest node last. This way cluster singletons are only moved once during a full rolling upgrade. Otherwise, in the worst case cluster singletons may be migrated from node to node which requires coordination and initialization overhead several times.

and it might introduce unnecessary delays.

An optional optimization is to leave the oldest node running a `ClusterSingleton` until last
to avoid it having to move more than once. 

> ## Migrating Untyped to Typed

Reviewer 2: ```suggestion
## Migrating from Classic to Typed Actors
```

Reviewer 2: What are the challenges here?

> It is recommended with a two step approach:

Reviewer 2: ```suggestion
A two step approach is recommended:
```

> * Deploy with the new nodes set to `akka.cluster.configuration-compatibility-check.enforce-on-join = off`

Reviewer 2: It's unclear to me how these steps relate to the migration from classic to typed actors?

and ensure all nodes are in this state
* Deploy again and with the new nodes set to `akka.cluster.configuration-compatibility-check.enforce-on-join = on`. 

The configuration from existing nodes should pass the @ref:[Cluster Configuration Compatibility Checks](#cluster-configuration-compatibility-check).

> Find out more about coexisting and @ref:[untyped to typed](../typed/coexisting.md#untyped-to-typed). 

Reviewer 2: ```suggestion
Find out more about classic and typed actors coexisting and the migration from @ref:[classic to typed actors](../typed/coexisting.md#untyped-to-typed).
```

### With Cluster Sharding and Persistence

> Rolling upgrades where shards on old nodes are running untyped persistent actors 

Reviewer 2: What does that mean exactly? Is that about migrating persistent entities from classic to typed? Or that on new nodes you could have new entities written with typed style?

Author: Removing it for adding when samples are done, but this is something user verified https://github.com/akka/akka/issues/26718

and new ones are running typed persistent behaviors have been tested successfully by the team and users.
Samples coming soon.

## Cluster Shutdown
 
@ref:[Coordinated Shutdown](../actors.md#coordinated-shutdown) will automatically run on SIGTERM when the cluster node sees itself as Exiting.

> Thus running shutdown tasks in a JVM shutdown hook is not recommended.

Reviewer 2: Why not? What kind of shutdown tasks and what is the recommendation instead?

Author: I was thinking the two links (Coordinated Shutdown, Graceful Shutdown) were enough? Do you have a suggestion or re-org that is clearer?

Author: Perhaps:

`Thus running shutdown tasks in a JVM shutdown hook is not recommended over Coordinated Shutdown`.

@ref:[Graceful shutdown](../cluster-sharding.md#graceful-shutdown) of Cluster Singletons and Cluster Sharding similarly happen automatically.
 

> In case of network failures it may still be necessary to set the node’s status to Down in order to complete the removal. 

Reviewer 2: ```suggestion
In case of network failures it may still be necessary to set the node's status to Down in order to complete the removal. 
```

Reviewer 2: I think we should expand here about when (or when not) and why SBR is needed.

Author: There is a link to SBR, Cluster Downing and Cluster Bootstrap just below that line.

Find out more about
* @ref:[Cluster Downing](../cluster-usage.md#downing) and providers
* [Cluster Bootstrap](https://doc.akka.io/docs/akka-management/current/bootstrap/index.html#rolling-updates) and Rolling updates
* [Split Brain Resolver](https://doc.akka.io/docs/akka-enhancements/current/split-brain-resolver.html)

## Cluster Configuration Compatibility Checks

> Relevant information on rolling updates and enforcing @ref:[Akka Cluster configuration compatibility checks](../cluster-usage.md#configuration-compatibility-check)

Reviewer 2: This is linked above as well, can the information be merged / expanded?

Author: Yes, done :)

parallel during the update, for example in blue green deployments.

For rolling updates related to Akka dependency version upgrades and the migration guides, please see
@ref:[Rolling Updates and Akka versions](../project/rolling-update.md)

#### This document covers: 
* [Serialization Compatibility](#serialization-compatibility)
* [Cluster Sharding](#cluster-sharding)
* [Migrating Untyped to Typed](#migrating-untyped-to-typed)
* [Cluster Shutdown](#cluster-shutdown)
* [Cluster Configuration Compatibility Check](#cluster-configuration-compatibility-check)
 
## Serialization Compatibility

There are two parts of Akka that need careful consideration when performing an rolling update.

1. Compatibility of remote message protocols. Old nodes may send messages to new nodes and vice versa.
1. Serialization format of persisted events and snapshots. New nodes must be able to read old data, and
   during the update old nodes must be able to read data stored by new nodes.

There are many more application specific aspects. It's important to have a strategy for serialization
format that can be evolved and you can find advice in

> @ref:[Persistence - Schema Evolution](../persistence-schema-evolution.md), which also applies to

Reviewer: I wonder if this is too strong for some (common) scenario's: for many applications, it might be fine to have some message loss during a rolling upgrade, as long as the application recovers afterwards. Of course that's then likely not strictly '0-impact', but still much better than a 'shut down and restart' upgrade, especially when there's typically no protocol changes.

Does a serialization failure currently just drop the message, or break the whole connection? If the latter, we might want to introduce an option to just drop the message instead (even though that's not without risks of course).

Anyway, just food for thought, not something that needs to go into this PR

Author: Good finer point to raise. I agree it's more detailed than may belong here for rolling updates, and this doc isn't about guarantees, more about recommendations and things to be aware of. And from there, finer details can ensue. Should we create a `discuss` issue for it?

Author: In https://doc.akka.io/docs/akka/current/serialization.html#serializer-with-string-manifest:

For rolling upgrades if the manifest is unknown/serialization failure:
* If the recommended java.io.NotSerializableException is thrown in `fromBinary` it is treated as a transient problem in the TCP remoting layer, the issue logged and message is dropped 
* Breaks the connection if other exceptions thrown: will tear down the TCP connection because it can be an indication of corrupt bytes from the underlying transport

remote messages when deploying with rolling updates.

### Retiring a serializer without downtime

One approach is carried out in two phases: 

1. Deploy a version of your application that still supports deserializing messages in the old form 
(your deserializer is still in "akka.actor.serializers"-section of your configuration, but it no longer serializes new messages 

> (it is removed from “akka.actor.serialization-bindings”-section). See @ref:[Serialization Configuration](../serialization.md#configuration) for details.

Reviewer: Shouldn't you first deploy a version that can deserialize the new format, but won't use it yet? Otherwise you're in trouble when a new node (with the new serializer) sends a message to an old node (that does not yet have the new deserializer).

1. After this upgrade has succeeded and there are no longer systems sending messages serialized with the decommissioned serializer, 
you can upgrade to a version that also no longer contains this serializer in `akka.actor.serializers`.

## Cluster Sharding

During a rolling upgrade, sharded entities receiving traffic may be moved during @ref:[shard rebalancing](../cluster-sharding.md#shard-rebalancing), 
to an old or new node in the cluster, based on the pluggable allocation strategy and settings.
When an old node is stopped the shards that were running on it may be allocated to one of the
other old nodes remaining in the cluster. 
 
It is recommended to leave the oldest node until last, because the shard coordinator 

> is running on that node. See @ref:[Cluster Shutdown](#cluster-shutdown) below for more detail.

Reviewer: I wonder if 'it is recommended' is too strong here: it definitely can be a useful optimization, but for many systems it shouldn't be an issue?

Author: Sure, I got that from something of ours that used the phrase, "it is recommended..." for this ;) Updated.

There are some cases when @ref:[a full cluster restart is needed](../cluster-sharding.md#rolling-upgrades).

## Migrating Untyped to Typed

It is recommended with a two step approach:

* Deploy with the new nodes set to `akka.cluster.configuration-compatibility-check.enforce-on-join = off`
and ensure all nodes are in this state
* Deploy again and with the new nodes set to `akka.cluster.configuration-compatibility-check.enforce-on-join = on`. 

The configuration from existing nodes should pass the @ref:[Cluster Configuration Compatibility Checks](#cluster-configuration-compatibility-check).
Find out more about coexisting and @ref:[untyped to typed](../typed/coexisting.md#untyped-to-typed). 

> ### With Cluster Sharding and Persistence

Author: This section related to and could be updated with #27342. And samples from @patriknw's work in akka/akka-samples#110 linked when merged.

Rolling upgrades where shards on old nodes are running untyped persistent actors 
and new ones are running typed persistent behaviors have been tested successfully by the team and users.
Samples coming soon.

## Cluster Shutdown
 
@ref:[Coordinated Shutdown](../actors.md#coordinated-shutdown) will automatically run when the cluster node sees itself as Exiting.
@ref:[Graceful shutdown](../cluster-sharding.md#graceful-shutdown) of Cluster Singletons and Cluster Sharding similarly happen automatically.
 

> But in case of network failures during rolling restarts and updates it 

Reviewer: ```suggestion
In case of network failures it 
```

may still be necessary to set the node’s status to Down in order to complete the removal. 

Find out more about
* @ref:[Cluster Downing](../cluster-usage.md#downing) and providers

> * [Cluster Bootstrap](https://doc.akka.io/docs/akka-management/current/bootstrap/index.md#rolling-updates) and Rolling updates

Reviewer: Perhaps also plug split brain resolver here?

> ### Cluster Shutdown Order

Reviewer: This is not specific to typed and already mentioned above, so perhaps this subsection is not needed?

Otherwise, in the worst case cluster singletons may be migrated from node to node which requires coordination and initialization 
overhead several times.

## Cluster Shutdown
 
@ref:[Coordinated Shutdown](../actors.md#coordinated-shutdown) will automatically run on SIGTERM when the cluster node sees itself as Exiting.
Thus running shutdown tasks in a JVM shutdown hook is not recommended over Coordinated Shutdown.
@ref:[Graceful shutdown](../cluster-sharding.md#graceful-shutdown) of Cluster Singletons and Cluster Sharding similarly happen automatically.
 
In case of network failures it may still be necessary to set the node's status to Down in order to complete the removal. 
@ref:[Cluster Downing](../cluster-usage.md#downing) details downing nodes and downing providers, and how using
[Split Brain Resolver](https://doc.akka.io/docs/akka-enhancements/current/split-brain-resolver.html) can help resolve
split brain during network partitions.

Additionally, [Cluster Bootstrap](https://doc.akka.io/docs/akka-management/current/bootstrap/index.html#rolling-updates)

> can be leveraged during rolling updates for joining and downing nodes in the cluster.

Author: @jrudolph yes there is quite a lot of information here related to rolls https://doc.akka.io/docs/akka-management/current/bootstrap/index.html#rolling-updates

> can be leveraged during rolling updates for joining and downing nodes in the cluster.

Author: Clarified and pushed if you can take another look, LMK what you think.
